### PR TITLE
coding-style.md: add various rules for writing C expressions - r2

### PR DESCRIPTION
**Pull request by DDvO**
This is a revision of #39, which failed to pass the voting due to partly ambiguous texts and late comments about this.

Note that the fixup commit https://github.com/openssl/technical-policies/commit/254439f9322751cf9f626b54bd54375ac8aef02c contains the two changes proposed today.

### CODE REVIEW

File: policies/coding-style.md [hunk 1]
```diff
@@ -648,6 +650,10 @@ but do this instead:
     if (strcmp(a, "FOO") == 0)
     if ((p = BN_new()) == NULL)
 ```
+Boolean values shall be used directly as usual, e.g.,
+```
```
Comment (levitte) [inline policies/coding-style.md:654]: ` ``` c `, please

Comment (DDvO) [reply policies/coding-style.md]: Oops, done.

File: policies/coding-style.md [hunk 2]
```diff
@@ -584,28 +594,99 @@ less complicated than coding pure assembly, and it doesn't provide the
...
+For example, do not write
+```c
+    if ((p == NULL) && (!f(((2 * x) + y) == (z++))))
+```
+but
+```c
+    if (p == NULL && !f(2 * x + y == z++)).
+```
 
+For clarity, always put parenthese
```
Comment (paulidale) [inline policies/coding-style.md:616]: _parenthese_ ? _parentheses_ !

File: policies/coding-style.md [hunk 3]
```diff
@@ -584,28 +594,99 @@ less complicated than coding pure assembly, and it doesn't provide the
...
+when mixing the logical `&&` and `||` operators,
+mixing comparison operators like `<=` and `==,`
+or mixing bitwise operators like  `&` and `|` .
```
Comment (paulidale) [inline policies/coding-style.md:619]: errant double space before \`&\`

File: policies/coding-style.md [hunk 4]
```diff
@@ -584,28 +594,99 @@ less complicated than coding pure assembly, and it doesn't provide the
...
+    if (i != 0)
+    if ((x & MASK) == 0)
+    if (strcmp(a, "FOO") == 0)
+    if ((p = BN_new()) == NULL)
+```
+Boolean values shall be used directly as usual, e.g.,
+```c
+if (check(x) && !success(y))
+```
+
```
Comment (paulidale) [inline policies/coding-style.md:657]: What about a note: _Many functions can return 0 or a negative value on error and the Boolean forms need to be used with care._? This is a very common error in the code base.

**Review by paulidale — CHANGES_REQUESTED** [review]

File: policies/coding-style.md (hunk 2 above)
Comment (DDvO) [reply policies/coding-style.md]: Fixed

File: policies/coding-style.md (hunk 3 above)
Comment (DDvO) [reply policies/coding-style.md]: Double space in Markdown should not make a difference, but removed anyway.

File: policies/coding-style.md (hunk 4 above)
Comment (DDvO) [reply policies/coding-style.md:659]: Yeah, good to add such a caveat - done.

**Review by paulidale — APPROVED** [review]

File: policies/coding-style.md (hunk 2 above)
Comment (t8m) [reply policies/coding-style.md]: @DDvO You did not push the fix here.

File: policies/coding-style.md [hunk 5]
```diff
@@ -584,28 +594,101 @@ less complicated than coding pure assembly, and it doesn't provide the
...
+    if ((p == NULL) && (!f(((2 * x) + y) == (z++))))
+```
+but
+```c
+    if (p == NULL && !f(2 * x + y == z++)).
+```
 
+For clarity, always put parentheses
+when mixing the logical `&&` and `||` operators,
+mixing comparison operators like `<=` and `==,`
```
Comment (t8m) [inline policies/coding-style.md:618]: `==`, not `==,`

**Review by t8m — CHANGES_REQUESTED** [review]
One typo

File: policies/coding-style.md (hunk 5 above)
Comment (DDvO) [reply policies/coding-style.md]: Good  spotting - fixed.

**Review by t8m — APPROVED** [review]

**Review by mattcaswell — APPROVED** [review]

**Comment by DDvO** [comment]
Thanks for the further discussion and approvals, also regarding the related PRs #47 and #48.

What does the formal process require now?
Has the new 14-days waiting period already started on them after which the OTC votings can be (re-)done?

**Comment by t8m** [comment]
> Has the new 14-days waiting period already started on them, after which the OTC votings can be (re-)done?

Yes, that period starts when the PR is submitted.

**Review by slontis — APPROVED** [review]

**Comment by t8m** [comment]
Starting the vote for coding-style.md: add various rules for writing C expressions - r2 at commit [2dc5c2b](https://github.com/openssl/technical-policies/pull/46/commits/2dc5c2b3330ab4211526cc9679ca0707cc91ad02)

**Comment by t8m** [comment]
Vote: [+1]

**Comment by paulidale** [comment]
Vote: [+1]

**Comment by beldmit** [comment]
Vote: [+1]

**Comment by mattcaswell** [comment]
Vote: [+1]

**Comment by slontis** [comment]
Vote: [+1]

**Comment by romen** [comment]
Vote: [+0]

**Comment by kroeckx** [comment]
Voting +1

**Comment by levitte** [comment]
Vote: [+1]

**Pull request closed without merging**